fhir-database-utils/src/main/java/com/ibm/fhir/database/utils/query/WhereClause.java

```diff
@@ -1,34 +1,78 @@
 /*
- * (C) Copyright IBM Corp. 2019
+ * (C) Copyright IBM Corp. 2019,2021
  *
  * SPDX-License-Identifier: Apache-2.0
  */
 
 package com.ibm.fhir.database.utils.query;
 
-import java.util.ArrayList;
-import java.util.List;
-import java.util.StringJoiner;
+import com.ibm.fhir.database.utils.query.expression.DebugExpNodeVisitor;
+import com.ibm.fhir.database.utils.query.expression.StringExpNodeVisitor;
+import com.ibm.fhir.database.utils.query.node.ExpNode;
+import com.ibm.fhir.database.utils.query.node.ExpNodeVisitor;
+import com.ibm.fhir.database.utils.query.node.PredicateParser;
 
 /**
  * The WhereClause SQL definition
  */
 public class WhereClause {
-    private List<String> predicates = new ArrayList<>();
+    private final PredicateParser predicateParser;
 
-    public void addPredicate(String predicate) {
-        this.predicates.add(predicate);
+    public WhereClause() {
+        this.predicateParser = new PredicateParser();
+    }
+
+    /**
+     * @return true if this where clause does not contain any predicates
+     */
+    public boolean isEmpty() {
+        return this.predicateParser.isEmpty();
+    }
+
+    /**
+     * Getter for the predicateParser owned by this WhereClause
+     * @return
+     */
+    public PredicateParser getPredicateParser() {
+        return this.predicateParser;
     }
 
     @Override
     public String toString() {
-        StringBuilder whereClauseBuilder = new StringBuilder();
-        whereClauseBuilder.append("WHERE ");
-        StringJoiner joiner = new StringJoiner(" ");
-        for (String predicate : predicates) {
-            joiner.add(predicate);
+        StringBuilder result = new StringBuilder();
+
+        if (!isEmpty()) {
+            // Render the predicate expression tree as a string
+            ExpNode predicate = predicateParser.parse();
+            StringExpNodeVisitor visitor = new StringExpNodeVisitor();
+            result.append("WHERE ");
```

Comment:
use the sql constants class?

Reply:
Constants to be revisited when we remove the old query builder.